File: flutter_app/lib/store/app_reducer.dart
import 'package:flutter_app/store/actions/set_sensors_action.dart';
import 'package:flutter_app/store/actions/update_sensor_action.dart';
import 'package:flutter_app/store/app_state.dart';

class AppReducer {
  static AppState Execute(AppState state, dynamic action) {
    state.lastAction = action;
    if (action is SetSensorsAction) {
      state.sensors = (action as SetSensorsAction).payload;
      return state;
    }
    if (action is UpdateSensorAction) {
      var updateAction = action as UpdateSensorAction;
      var updateSensor =
          state.sensors.firstWhere((f) => f.id == updateAction.sensorId);
      if (updateSensor == null) {
        return state;
      }
      updateSensor.latestPollutionLevel = updateAction.pollutionLevel;
      updateSensor.readings.add(updateAction.payload);
      updateSensor.readings.sort((a, b) =>
          (b.created.difference(a.created).inMilliseconds) > 0 ? 1 : 0);
      updateSensor.readings = updateSensor.readings.take(10).toList();
      return state;
    }
    return state;
  }
}

File: src/Clients/mobile/Mobile.Core/Store/AppReducer.cs

using Mobile.Core.Store.Actions;
using Redux;
using System.Linq;

namespace Mobile.Core.Store
{
    public static class AppReducer
    {
        public static AppState Execute(AppState state, IAction action)
        {
            var newState = (AppState)state.Clone();
            newState.LastAction = action;
            if (action is SetFilterParameterAction)
            {
                newState.FilterByParameter = (action as SetFilterParameterAction).Payload;
                return newState;
            }
            if (action is SetSensorsAction)
            {
                newState.Sensors = (action as SetSensorsAction).Payload;
                return newState;
            }
            if (action is UpdateSensorAction)
            {
                var updateAction = action as UpdateSensorAction;
                var updateSensor = newState.Sensors.FirstOrDefault(f => f.Id == updateAction.SensorId);
                if (updateSensor == null)
                {
                    //throw new ArgumentNullException($"No such sensor to update {nameof(updateSensor)}");
                    return newState;
                }
                updateSensor.LatestPollutionLevel = updateAction.PollutionLevel;
                updateSensor.Readings.Add(updateAction.Payload);
                updateSensor.Readings = updateSensor.Readings.OrderByDescending(f => f.Created).Take(10).ToList();
                return newState;
            }
            return newState;
        }
    }
}
